lib/saml/user_attributes/mhv.rb

```diff
@@ -77,10 +86,50 @@ def loa_highest
         PREMIUM_LOAS.include?(account_type) ? 3 : (idme_loa || loa_current)
       end
 
-      def loa_highest_available
-        3
+      # Attributes from MVI
+      def first_name
+        mvi&.profile&.given_names&.first
+      end
+
+      def middle_name
+        mvi&.profile&.given_names&.last
+      end
+
+      def last_name
+        mvi&.profile&.family_name
+      end
+
+      def birth_date
+        mvi&.profile&.birth_date
+      end
+
+      def ssn
+        mvi&.profile&.ssn
+      end
+
+      def gender
+        mvi&.profile&.gender
+      end
+
+      # Probably need to rescue from when ICN query returns no result returning NullMVI
+      # Logging these various scenarios would provide useful data
+      def mvi
+        @mvi ||= begin
+          if PREMIUM_LOAS.include?(account_type)
+            if mhv_icn.present?
+              # What if the ICN doesn't return a hit when querying MVI???
```

Comment:
@saneshark does this happen? no profile returned when querying MVI by ICN? A bit confused around the need for `NullMvi`. Is it the user that can't be persisted? I think we should revisit those user validations and possibly move them as a guard against querying MVI. Probably need more context behind everything here 😄  but it's getting hard to follow with a specialized null class and some things coming from SAML and methods hitting MVI which mirroring methods in the user class `mvi&.profile&.gender` etc.

Reply:
@kreek so the problem is that for MHV sign-in we don't have first_name, last_name, dob, gender, ssn, etc. So to get those values I need to side-load MVI so that I can hydrate current_user with the required attributes for LOA3, see validations on `app/models/user.rb.` 

It's true, we talked about the fact that maybe those validations could just go away, but I'm not entirely sure what the ramifications are, and furthermore, I'm concerned with how the existing exception handling on mvi in current user bubbles up to the user. By fetching MVI pre-emptively in the saml_callback I can ensure that the user is not persisted, but doing so in current_user itself would mean that the user is already persisted. And yes, an ICN value could possibly not be provided, additionally, the ICN that is provided might not resolve to an actual MVI record if what MHV has is outdated. Finally, there is no guarantee that MVI service is available, it could be down. In each of these three scenarios we need errors that are unique to the MHV sign-in experience so that we can communicate remediation to the user effectively, otherwise we're left with a very generic, you cannot login (even though your credentials are good and you're a premium MHV user). The NullMVI object is something I'm more or less just playing with here to see if it is a better pattern. It was the first solution to the problem that occurred to me. But I think it helps illustrate the problem at hand.

To be clear, I'm really not happy with this proposed implementation, but it's the least heavy handed approach I could think of until we can address authorization and current user hydration. The way mvi is delegated on current_user is a challenge to deal with currently and I didn't think I would have enough time to refactor it.

Reply:
And yes, you're right. I have not yet addressed the edge cases or introduced the granular error handling I'm describing for the 3 scenarios in which a user is logged in but necessary attributes are unavailable. I was going to address those in a separate PR.

Reply:
Another approach, one that I personally would advocate for, but haven't had time to explore... would be for MVI to ALWAYS be loaded in the saml callback when the user is "loa3" or "premium." and that current_user no longer needs to delegate like it does today, but it still doesn't address one important problem in that you could be signed in as an identity proofed user and not have ssn, dob, etc (should you still be characterized as LOA3 seeing as certain services authorize with the expectation of these attributes being available?)

Reply:
Maybe this will be helpful, maybe not... but could be a starting point for communicating how to simplify things.

<img width="965" alt="screen shot 2017-10-17 at 12 35 21 am" src="https://user-images.githubusercontent.com/591737/31648401-166dffb2-b2d3-11e7-8b4e-c327fc31f0e0.png">

Reply:
@saneshark Thanks for that definitely sheds light on the complexity behind it. Now just need to design it so that complexity is hidden from the classes that work with authentication and authorization.

Stepping back a bit there’s really only three things going on here:
1. a user logs in from somewhere (authentication)
2. their VA profile is retrieved
3. their permissions are checked (authorization)

2 and 3 right now are a bit too intertwined in our current codebase. If we can have thwo distinct classes that act as an API/facade for profile and permissions it could simplyfy things. Then you can pipe things together Login | Profile | Permissions. 

Login, as painful as it was, is pretty much solved. A user logs in and they are in some state. That user then gets passed as input to VA Profile class to be filled out either by querying by full attributes or icn. The user that come out of that ‘pipe’ is now in another state. They want to interact with a service. The user and and the service are passed into the Permissions class which checks a policy and get a yay or nay.

MHV (and sometimes DSLogon?) seem to be here to throw a wrench into the above being simple 😀  For ID.me users MVI is not authentication, it’s just data (that’s used for authorization), but for the other two it’s authorization and authentication. Ideally MVI would never be used for authentication, despite what we were intially told it is not always the source of truth, it has duplicate records, it can be out of date, MHV ids might not line up. 

Can we make MVI irrelevant to authentication? A user still goes through the Login and Profile pipes and comes out, if they couldn’t hit MVI then the Permissions pipe always rejects them. Is that ok?

Reply:
The pipeline described by @kreek makes a lot of sense. 

To complicate matters I think there are some site features for which if MVI is not reachable we still want to allow certain actions. But I agree that authentication and profile are inputs (sometimes nil inputs!) to the authorization decision, and the authorization is where the complex business rules need to live.

Reply:
@kreek I'm going to try to make MVI not required as part of authentication. For a non-premium user it's absolutely fine I think because the navbar on the front-end could just show the ID.me wallet email address instead of first name.  The problem arose for a premium MHV user, without first name, we wouldn't have name and other important attributes until we query MVI. The intermediary abstraction does make a lot of sense in helping to support this, but of course it touches upon a bit more refactoring than I felt comfortable making at the time.